### PR TITLE
WIP: gnrc/rpl mrhof

### PR DESCRIPTION
**Pull request by bergzand**
This WIP is an initial work on mrhof ([rfc 6719](https://tools.ietf.org/html/rfc6719)) for RPL.

Depends on #6873 (& co) for ETX statistics.

todo:
- [x] Decent link local ipv6-address to L2-address conversion
- [x] Parent selection (sec. 3.2)
- [x] Preferred parent selection (sec 3.2.2) 
- [x] Correct rank computation (sec. 3.2).
~~- [ ] metric containers~~
- [x] MAX_LINK_METRIC
- [x] MAX_PATH_COST

### CODE REVIEW

**Comment by bergzand** [comment]
Now depends on #7468 for parent set calculation.

**Comment by bergzand** [comment]
Should be feature complete now, so testers are welcome. I'm not going to implement metric containers in this PR and only support ETX as a metric.

PR dependency graph might have gotten a bit messed up, but at least #6873 for the neighbor statistics and #7449 for a bit of additional ETX output should be enough.

**Comment by bergzand** [comment]
As a remark when people are going to test this. This PR relies heavily on ETX functionality of the netstats neighbor module. For this to work correctly, support from the radio is really a must. Only succes/noack reporting is not sufficient to really estimate the link quality. As far as I know only the mrf24j40 and the at86r233 support reading of transmission retries needed for accurate ETX estimation. For the mrf24j40 I have a patch ready contained in the chain of commits in this PR. I don't have access to an at86rf233, so I can't test any patch for transmission retries for that device.

So for testing, I would either advise using mrf24j40's or write a patch for the at86rf233 and use that one.
There are probably more radios that support retransmission statistics, so if #7448 is pulled, similar PR's are welcome :)

**Comment by bergzand** [comment]
As of now, this PR would cause gnrc_rpl to depend on the netstats_neighbor module from #6873. Would it be better to make a gnrc_rpl_mrhof submodule such that the mrhof code (and dependencies) becomes optional? I can imagine lightweight nodes not wanting the extra memory required by the netstats_neighbor module.

**Comment by bergzand** [comment]
Rebased and adjusted for current master

**Comment by bergzand** [comment]
Rebased and uncrustified

File: sys/include/net/netstats/neighbor.h [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * @return ptr to the record
+ * @return NULL if no records remain
+ */
+netstats_nb_t *netstats_nb_get_next(netstats_nb_t *first, netstats_nb_t *prev);
+
+/**
+ * @brief Store this neighbor as next in the transmission queue.
+ *
+ * Set len to zero if a nop record is needed, for example if the
```
Comment (cgundogan) [inline sys/include/net/netstats/neighbor.h]: `len` =>  `@p len`

File: sys/include/net/netstats/neighbor.h [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * @return ptr to the record
+ */
+netstats_nb_t *netstats_nb_get_recorded(netdev_t *dev);
+
+/**
+ * @brief Update the next recorded neighbor with the provided numbers
+ *
+ * This only increments the statistics if the length of the l2-address of the retrieved record
+ * is non-zero. see also @ref netstats_nb_record. The numbers indicate the number of transmissions
```
Comment (cgundogan) [inline sys/include/net/netstats/neighbor.h]: ~~s~~See

File: sys/net/gnrc/link_layer/netdev/gnrc_netdev_eth.c [hunk 3]
```diff
@@ -218,9 +222,17 @@ static int _send(gnrc_netdev_t *gnrc_netdev, gnrc_pktsnip_t *pkt)
         if ((netif_hdr->flags & GNRC_NETIF_HDR_FLAGS_BROADCAST) ||
             (netif_hdr->flags & GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
             gnrc_netdev->dev->stats.tx_mcast_count++;
+#ifdef MODULE_NETSTATS_NEIGHBOR
+            DEBUG("l2 stats: Destination is multicast or unicast, NULL recorded");
```
Comment (cgundogan) [inline sys/net/gnrc/link_layer/netdev/gnrc_netdev_eth.c]: are you missing a `\n` here?

File: sys/net/gnrc/link_layer/netdev/gnrc_netdev_ieee802154.c [hunk 4]
```diff
@@ -216,9 +220,17 @@ static int _send(gnrc_netdev_t *gnrc_netdev, gnrc_pktsnip_t *pkt)
     if (netif_hdr->flags &
         (GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
             gnrc_netdev->dev->stats.tx_mcast_count++;
+#ifdef MODULE_NETSTATS_NEIGHBOR
+            DEBUG("l2 stats: Destination is multicast or unicast, NULL recorded");
```
Comment (cgundogan) [inline sys/net/gnrc/link_layer/netdev/gnrc_netdev_ieee802154.c]: also, `\n` here?

File: sys/net/netstats/netstats_neighbor.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+ */
+
+#include <errno.h>
+
+#include "net/netdev.h"
+#include "net/netstats/neighbor.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+//static char addr_str[30];
```
Comment (cgundogan) [inline sys/net/netstats/netstats_neighbor.c]: can be removed?

File: sys/net/gnrc/link_layer/netdev/gnrc_netdev_eth.c (hunk 3 above)
Comment (bergzand) [reply sys/net/gnrc/link_layer/netdev/gnrc_netdev_eth.c]: Yes :(

File: sys/include/net/netstats/neighbor.h (hunk 2 above)
Comment (bergzand) [reply sys/include/net/netstats/neighbor.h]: Ack, will fix

File: sys/include/net/netstats/neighbor.h (hunk 1 above)
Comment (bergzand) [reply sys/include/net/netstats/neighbor.h]: Ack, will fix

File: sys/net/netstats/netstats_neighbor.c (hunk 5 above)
Comment (bergzand) [reply sys/net/netstats/netstats_neighbor.c]: Whoops, yes

**Comment by bergzand** [comment]
@cgundogan Thanks for the (initial) review. I noticed that this PR was missing the "waiting for" tag. I've addressed your comments in #6873.

**Pull request closed without merging**